cpp/ql/src/experimental/Security/CWE/CWE-020/NoCheckBeforeUnsafePutUser.ql

```diff
@@ -16,44 +16,143 @@
 
 import cpp
 import semmle.code.cpp.dataflow.DataFlow
+import semmle.code.cpp.dataflow.TaintTracking
+import DataFlow::PathGraph
 
-class WriteAccessCheckMacro extends Macro {
+/*
+ * using functions' name to model linux syscall which usually
+ * starts with prefix like:
+ *   C_SYSC_
+ *   SYSC_
+ *   compat_Sys_
+ *   Sys_
+ */
+
+class SysCallFunction extends Function {
+  SysCallFunction() {
+    this.getName().regexpMatch("C_SYSC_[a-zA-Z]+") or
+    this.getName().regexpMatch("SYSC_[a-zA-Z]+") or
+    this.getName().regexpMatch("compat_SyS_[a-zA-Z]+") or
+    this.getName().regexpMatch("SyS_[a-zA-Z]+")
```

Comment:
> Is there a reason for using the `[a-zA-Z]+` pattern rather than `.*` (or `.match("C_SYSC_%")`)? I'm not saying one or other approach is better, just curious about the intended behaviour e.g. for functions potentially with numeric digits in their names.

Hello @geoffw0 ,

As I commented in the query SysCallFunction is used to model linux syscall, and as far as I know there is no linux syscall have numeric digits in their names. And frankly speaking I don't think have numeric digits in function names is a good coding style hahah:)

Reply:
Per https://filippo.io/linux-syscall-table/ there seem to be some syscalls with digits such as pread64, dup2, etc.

Reply:
> Per https://filippo.io/linux-syscall-table/ there seem to be some syscalls with digits such as pread64, dup2, etc.

Sorry for missing your reply, Thank you. I will try out another way to model linux syscall :)